File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.diameter = 0

    def get_diameter(self, node):
        l = self.get_diameter(node.left) if node.left else 0
        r = self.get_diameter(node.right) if node.right else 0

        self.diameter = (l + r) if (l + r) > self.diameter else self.diameter

        return 1 + (l if l > r else r)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.get_diameter(root)
        return self.diameter
        

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int len = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        dfs(root);
        return len;    
    }

    public int dfs(TreeNode node) {
        if (node == null) return 0;

        int left = dfs(node.left);
        int right = dfs(node.right);

        len = Math.max(left + right, len);
        return Math.max(left, right) + 1;
    }
}